Implement `get_*_draw_list` and `DrawList` drawing functions

tangmi opened the pull request:
Closes #220 (when 0.1 lands).

`ImDrawList_Get*DrawList` are very powerful functions and this PR exposes the capability in Rust.

- Uses `0.1`'s `AsRef<str>`-based way of passing in text.
- Uses `ImColor` from: https://github.com/Gekkio/imgui-rs/blob/1f5bb7c449e9f6f13dfd84640c426e7cdd91be49/src/window_draw_list.rs#L17-L53
- Does not implement the "advanced" (callbacks) or internal helper APIs
- Uses `ImVec2/4` and adds a conversion from `ImVec2/4_Simple` to `ImVec2/4`

hrydgard commented:
Cool! Looks great, exactly what I'll need. Any idea what the schedule for 0.1 is?

tangmi commented:
I think that's up to @Gekkio, the relevant links are #167 and #168.

tangmi commented:
Whoops, I just found https://github.com/Gekkio/imgui-rs/blob/master/src/window_draw_list.rs, which probably this probably should share code with.

hrydgard commented:
Right, indeed..

tangmi commented:
@Gekkio I've cleaned this up, added an example to `imgui-examples`, and squashed the changes into a single commit. This should be ready to review whenever you have time!

Gekkio commented:
Thanks! :smile: 

I didn't yet have a chance to do a full review, but here's some things I noticed quickly:

* this seems to target 0.1-dev, but that is a throwaway branch
* starting from 0.1, ImVec2/ImVec4 are not allowed in safe APIs
* what's the difference between this and the existing draw list API?
* what's the reason of having `ImColor`? Colours are `[f32; 4]` everywhere else
* all `get_xx_draw_list` functions cause UB, because they can generate references with any lifetime
* all `get_xx_draw_list` functions cause UB, because they can generate multiple mutable references to the same thing
* reusing the `DrawList` type for this use case causes UB, because you can have active references to idx_buffer/vtx_buffer/DrawCmdIterator while also modifying the data underneath

tangmi commented:
Thanks for the reply, I'll try and update the code to respond to this feedback, so please hold off on a full review for now!

>    this seems to target 0.1-dev, but that is a throwaway branch

I'll rebase this onto master

>    starting from 0.1, ImVec2/ImVec4 are not allowed in safe APIs

Sounds good, I'll change the usage to `[f32; 2]` and `[f32; 4]`

>    what's the difference between this and the existing draw list API?

No difference! I thought that `0.1-dev` was newer than master! I'll change this to use the existing draw list API in master.

>    what's the reason of having ImColor? Colours are [f32; 4] everywhere else

I change this to use `[f23; 4]`, although I think there's something to be said for having a type to document (and enforce) that colors stay within the [0, 1] range.

>    all get_xx_draw_list functions cause UB, because they can generate references with any lifetime

I'm still learning about unsafe Rust--I thought something like this implies `'ui: 'a`? (please correct me here!)

```
impl<'ui> Ui<'ui> {
    pub fn get_window_draw_list<'a>(&self) -> &'a mut DrawList {
        ...
    }
}
```

I think `DrawList` in `master` has a `'ui` lifetime, so I can have this function instead be

```
impl<'ui> Ui<'ui> {
    pub fn get_window_draw_list(&self) -> &'ui mut DrawList<'ui> {
        ...
    }
}
```

if that'll be better?

>    all get_xx_draw_list functions cause UB, because they can generate multiple mutable references to the same thing
>    reusing the DrawList type for this use case causes UB, because you can have active references to idx_buffer/vtx_buffer/DrawCmdIterator while also modifying the data underneath

I _think_ these can be addressed together. I'll try and find a solution for this. I initially had the `get_xx_draw_list` functions take `&mut self`, but I remember that causing API usage issues since one would often want other parts of `Ui` while doing a draw.

Gekkio commented:
> I'm still learning about unsafe Rust--I thought something like this implies 'ui: 'a? (please correct me here!)

I don't think it does...if we remove the effect of lifetime elision, I think it's basically equal to: 

```
impl<'ui> Ui<'ui> {
    pub fn get_window_draw_list<'self, 'a>(&'self self) -> &'a mut DrawList {
        ...
    }
}
```

The function takes a `&'self Ui<'ui>` (with some lifetime `'self` for the reference and `'ui` for the struct data) and returns a mutable reference `&'a mut DrawList` with *any lifetime 'a* (which could be `'static` and it would compile). I think the right thing in this case might be to drop the explicit lifetimes completely.

It's also worth noting that the current draw list API might have safety holes as well, so you shouldn't trust it without thinking...in 0.1 I've tried to fix many safety issues, but the window draw list API is one of the many things I didn't have time or energy to look at.

tangmi commented:
@Gekkio would you consider using bindgen (or checking in bindgen output, with instructions on how to regenerate)? I'm running into issues upgrading master to imgui >= 1.69, which is required for this PR. the 0.1-dev branch seems to use bindgen and be updated to 1.69.

tangmi commented:
Abandoning in favor of #242
